### PR TITLE
Fix Nupkg/Nuspec UTF-16 encoding and support uppercase dependencies

### PR DESCRIPTION
**Pull request by sverdlov93**
- [ ] All [tests](https://github.com/jfrog/build-info-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] All [static analysis checks](https://github.com/jfrog/build-info-go/actions/workflows/analysis.yml) passed.
- [ ] This pull request is on the dev branch.
- [ ] I used gofmt for formatting the code before submitting the pull request.
-----
When parsing Nuget `nuspec` files sometimes they can be utf16 files so we need to encode them because Golang's Unmashall doesn't support utf-16 encoding.
The previous fix wasn't enough for all cases.
Added a real utf16 example to test this.
Also, fixed an issue where Nuget dependencies with uppercase letters couldn't be resolved.

should fix https://github.com/jfrog/jfrog-azure-devops-extension/issues/475

### CODE REVIEW

File: build/utils/dotnet/dependencies/packagesconfig.go [hunk 1]
```diff
@@ -315,10 +316,21 @@ type group struct {
 func xmlUnmarshal(content []byte, obj interface{}) (err error) {
 	err = xml.Unmarshal(content, obj)
 	if err != nil {
-		// Sometimes the nuspec file is wrongly encoded in utf-16, the actual encoding is utf-8 but the xml header has 'encoding="utf-16"' key.
+		// Failed while trying to parse xml file. Nuspec file could be an utf-16 encoded file.
```
Comment (fourpastmidnight) [inline build/utils/dotnet/dependencies/packagesconfig.go:319]: Yes! This is the issue I was running into. Now I see why the original fix didn't work.

**Comment by fourpastmidnight** [comment]
What's going on with this PR, as I haven't seen any movement on this for about 2 weeks? I'm looking for this fix, as I'm gearing up to go to production soon and would like to know if I'll be able to use the JFrog Restore task in my build pipelines.

File: build/utils/dotnet/dependencies/packagesconfig.go [hunk 2]
```diff
@@ -315,10 +316,21 @@ type group struct {
...
+		// xml.Unmarshal doesn't support utf-16 encoding, so we need to decode the utf16 by ourselves.
+
+		buf := make([]uint16, len(content)/2)
+		for i := 0; i < len(content); i += 2 {
+			buf[i/2] = binary.LittleEndian.Uint16(content[i:])
+		}
+		// Remove utf-16 Byte Order Mark (BOM) if exists
+		utf16BOM := "\uFEFF"
```
Comment (yahavi) [inline build/utils/dotnet/dependencies/packagesconfig.go:327]: Let's make it a file const

File: build/utils/dotnet/dependencies/packagesconfig.go [hunk 3]
```diff
@@ -315,10 +316,21 @@ type group struct {
...
+		stringXml := strings.ReplaceAll(string(utf16.Decode(buf)), utf16BOM, "")
+
 		// xml.Unmarshal doesn't support utf-16 encoding, so we need to convert the header to utf-8.
-		utf8Bytes := bytes.Replace(content, []byte("utf-16"), []byte("utf-8"), 1)
-		err = xml.Unmarshal(utf8Bytes, obj)
+		stringXml = strings.Replace(stringXml, "utf-16", "utf-8", 1)
+
+		err = xml.Unmarshal([]byte(stringXml), obj)
```
Comment (yahavi) [inline build/utils/dotnet/dependencies/packagesconfig.go:333]: Let's add checkerr

File: build/utils/dotnet/dependencies/packagesconfig.go (hunk 1 above)
Comment (yahavi) [inline build/utils/dotnet/dependencies/packagesconfig.go:319]: Let's add a debug log message here to mention that we are using this fallback

**Review by yahavi — CHANGES_REQUESTED** [review]

File: build/utils/dotnet/dependencies/packagesconfig.go (hunk 3 above)
Comment (sverdlov93) [reply build/utils/dotnet/dependencies/packagesconfig.go:333]: @yahavi its build-info-go, no check error here

**Comment by fourpastmidnight** [comment]
Awesome, glad to see this has been completed. Has this been formally released?

**Comment by fourpastmidnight** [comment]
Did you guys actually test that this works with the package mentioned in issues #464 and #475? I'm _still_ getting an error trying to use the JFrog task to restore `microsoft.web.infrastructure-1.0.0`.

**Pull request merged**